core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -0,0 +1,1152 @@
+            if (value == null) {
+                value = e.value;
+            }
+            entries[i] = e.next;
+            e.remove();
+            --size;
+        }
+
+        return value;
+    }
+
+    private static boolean keyEquals(@Nullable AsciiString a, CharSequence b) {
+        return a != null && (a == b || a.contentEqualsIgnoreCase(b));
+    }
+
+    // Conversion functions
+
+    @Nullable
+    private static Integer toInteger(@Nullable String v) {
+        try {
+            return v != null ? Integer.parseInt(v) : null;
+        } catch (NumberFormatException ignore) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private static Long toLong(@Nullable String v) {
+        try {
+            return v != null ? Long.parseLong(v) : null;
+        } catch (NumberFormatException ignore) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private static Float toFloat(@Nullable String v) {
+        try {
+            return v != null ? Float.parseFloat(v) : null;
+        } catch (NumberFormatException ignore) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private static Double toDouble(@Nullable String v) {
+        try {
+            return v != null ? Double.parseDouble(v) : null;
+        } catch (NumberFormatException ignore) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private static Long toTimeMillis(@Nullable String v) {
+        if (v == null) {
+            return null;
+        }
+
+        try {
+            @SuppressWarnings("UseOfObsoleteDateTimeApi")
+            final Date date = DateFormatter.parseHttpDate(v);
+            return date != null ? date.getTime() : null;
+        } catch (RuntimeException ignore) {
+            return null;
+        }
+    }
+
+    private static String fromTimeMillis(long value) {
+        return StringValueConverter.INSTANCE.convertTimeMillis(value);
+    }
+
+    private static String fromObject(Object value) {
+        final String strVal = StringValueConverter.INSTANCE.convertObject(value);
+        assert strVal != null : value + " converted to null.";
+        return strVal;
+    }
+
+    // hashCode(), equals() and toString()
+
+    @Override
+    public final int hashCode() {
+        int result = HASH_CODE_SEED;
+        for (AsciiString name : names()) {
+            result = (result * 31 + name.hashCode()) * 31 + getAll(name).hashCode();
+        }
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (!(o instanceof HttpHeaderGetters)) {
+            return false;
+        }
+
+        final HttpHeaderGetters that = (HttpHeaderGetters) o;
+        if (isEndOfStream() != that.isEndOfStream() ||
+            size() != that.size()) {
+            return false;
+        }
+
+        for (AsciiString name : names()) {
+            if (!getAll(name).equals(that.getAll(name))) {
```

Comment:
Just curious. If we have two HttpHeaderBase instances as follows:
```
HttpHeaderBase headers1 = new HttpHeaderBase(16);
headers1.add("name", "value1");
headers1.add("name", "value2");
headers1.add("name", "value3");

HttpHeaderBase headers2 = new HttpHeaderBase(16);
headers2.add("name", "value3");
headers2.add("name", "value2");
headers2.add("name", "value1");
```
Should `headers1.equals(headers2)` return `true`? If so, the current code seems returning `false`.

Reply:
I think it should return `false` because the header value order matters when the header name is same. Added test cases to `HttpHeadersBase`.